[FIX] *: share URL on the correct website if possible

seb-odoo opened the pull request:
rebase ff 

opw-1921030

seb-odoo commented:
TODO need to write tests

seb-odoo commented:
Can be merged standalone, but would work much better with the fixes on #29943 (to also apply here)

seb-odoo commented:
Code improved following our discussion (making it more specific and not trying to guess based on company).

seb-odoo commented:
@qsm-odoo can you please check the tour on commit [FIX] website_sale: correctly set access token for cart recovery ?

qsm-odoo commented on addons/website_sale/static/src/js/website_sale_tour_shop_cart_recovery.js:
```diff
@@ -0,0 +1,104 @@
+odoo.define('website_sale.tour_shop_cart_recovery', function (require) {
+    'use strict';
```
No indentation in odoo.define ;)

seb-odoo replied:
Unfortunately...

qsm-odoo commented on addons/website_sale/static/src/js/website_sale_tour_shop_cart_recovery.js:
```diff
@@ -0,0 +1,104 @@
+        url: '/shop?search=' + productName,
+        wait_for: base.ready(),
+    },
+    [
```
Also check the convention used by other tours for indentation of steps

seb-odoo replied:
Do we enforce this one? All the other tours I have written in the past months use the format I used here.

qsm-odoo replied:
Well, when there are no clear conventions, I use the one which is used by most other cases if it is not too stupid... so here I would use it. But no, you are not obliged to.

qsm-odoo commented on addons/website_sale/static/src/js/website_sale_tour_shop_cart_recovery.js:
```diff
@@ -0,0 +1,104 @@
+    [
+        {
+            content: "select " + productName,
+            trigger: '.oe_product_cart a:containsExact("' + productName + '")',
```
Not sure it is a good idea to compute trigger and content using variables. In this case, this does not increase readability and if the step fails at some point, people will have a hard time finding the step `.oe_product_cart a:containsExact("Acoustic Bloc Screens")`

seb-odoo replied:
Good point, I'll change to help grepping.

qsm-odoo commented on addons/website_sale/static/src/js/website_sale_tour_shop_cart_recovery.js:
```diff
@@ -0,0 +1,104 @@
+            content: "wait for rpc to be done and logout",
+            trigger: '#shop_cart_recovery_rpc_done',
+            run: function () {
+                window.location.href = "/web/session/logout?redirect=/";
```
Maybe avoid that step by doing the redirection in the `then` above and adding an extra trigger on the next step ?

seb-odoo replied:
Good idea, don't even need an extra trigger because the "login" trigger is enough (the other trigger was for the end of `then`.

qsm-odoo commented on addons/website_sale/static/src/js/website_sale_tour_shop_cart_recovery.js:
```diff
@@ -0,0 +1,104 @@
+                $('.oe_login_form input[name="login"]').val("admin");
+                $('.oe_login_form input[name="password"]').val("admin");
+                $('.oe_login_form input[name="redirect"]').val(url);
+                $('.oe_login_form').submit();
```
Suppose you could save the .oe_login_form jQuery element in a variable

seb-odoo replied:
It was copied from another tour but ok, better with a variable.

seb-odoo commented:
I finished writing all the tests I wanted. Hopefully if I correctly fixed the runbot issue and it is finally green, then it can be reviewed and rebase-ff.

seb-odoo commented on addons/website/tests/test_get_current_website.py:
```diff
@@ -0,0 +1,104 @@
+class TestGetCurrentWebsite(HttpCase):
+
+    @api.multi
+    def _my_unlink(self):
```
self note: this can be removed

seb-odoo commented on addons/website/tests/test_get_current_website.py:
```diff
@@ -0,0 +1,104 @@
+
+
+@tagged('post_install', '-at_install')
+class TestGetCurrentWebsite(HttpCase):
```
self note: this can be transactioncase

seb-odoo commented:
runbot green, for me we can review now

seb-odoo commented:
daily reminder, this is relatively high in my merge priority

seb-odoo commented on odoo/tests/common.py:
```diff
@@ -530,6 +530,7 @@ def _chrome_start(self):
             '--user-data-dir': self.user_data_dir,
             '--disable-translate': '',
             '--window-size': '1366x768',
+            '--remote-debugging-address': HOST,
```
The goal of the change is that then it works fine even if the value of HOST is changed. We can change the config IP in only one place instead of several.

d-fence replied:
I agree

seb-odoo commented on odoo/tests/common.py:
```diff
@@ -572,7 +573,7 @@ def _json_command(self, command, timeout=3):
         while timeout > 0:
             try:
-                url = werkzeug.urls.url_join('http://127.0.0.1:%s/' % self.devtools_port, command)
+                url = werkzeug.urls.url_join('http://%s:%s/' % (HOST, self.devtools_port), command)
```
same reasoning here, store value in only one place.

seb-odoo commented on addons/website/tests/test_get_current_website.py:
```diff
@@ -0,0 +1,99 @@
+        website1.domain = website2.domain
+        self.assertEqual(Website._get_current_website_id('my2ndsite.com', country5.id), website1.id)
+
+        # CASE: overlaping domain: get exact match
```
overlapping

seb-odoo commented on addons/website/tests/test_get_current_website.py:
```diff
@@ -0,0 +1,99 @@
+        # CASE: domain set: get matching domain
+        self.assertEqual(Website._get_current_website_id('my-site-1.fr', False), website1.id)
+
+        # CASE: domain set: get matching domain (scheme and port supported)
```
add a test for the port too

seb-odoo commented:
The 2 first commits have been merged separately (see commit referenced by robodoo above).

Now I'm rebasing it and making my final own review.

seb-odoo commented:
Remaining points to discuss before review/merge:

- Do we actively try to match the correct port and/or scheme when selecting website based on domain? They can (or have to be) encoded in the field, but they are currently not used in the detection, just when creating links.
- `get_base_url` is currently redefined in several models because it is not a mixin (website_id specified separately on each of those models). We merge it this way in 12.0?

JKE-be commented:
1. if you can handle it in less than 2 lines (heu, yes so 1 line) in the filter yes.
2. no mixin in 12, maybe in master laterw e could reuse website mixin.

Probably on website helper that return (http_domain, domain) -> (scheme+(name+tld)+port, name+tld)

seb-odoo commented:
I added to take port into account when selecting website. This made me realize the previous code wasn't actually working as I wanted, so I fixed it too and added a test case for the port.

No change for mixin in 12.

Runbot green, for me we can review.

JKE-be commented on addons/website/static/src/js/tours/website_tour_reset_password.js:
```diff
@@ -0,0 +1,192 @@
+        },
+    },
+    {
+        content: "click on Admin",
```
for NEXT time,

since we dont want to test UI here, it will be better in one step to write on partner admin the correct website.
Here, if we change the webclient (td, widget class, automcomplete lib, ...)  your test will be broken, while your feature is still correct.

seb-odoo replied:
ok for next time

JKE-be commented on addons/website/tests/test_website_reset_password.py:
```diff
@@ -0,0 +1,33 @@
+            kwargs.update(force_send=False, raise_exception=False)
+            return original_send_mail(*args, **kwargs)
+
+        with patch.object(MailMail, 'unlink', lambda self: None), patch.object(MailTemplate, 'send_mail', my_send_mail):
```
??

why not simply send the email (that is btw catched by runbot in a mailctacher? )?

JKE-be replied:
ok, I know your futur answer, yes it is sent...

but why force_send=False, raise_exception=False is needed ?

JKE-be replied:
ok, I know you futur answer, yes it is explain

# we override send_mail because in HttpCase we don't have an smtp server
# without force_send and raise_exception, it is going to fail silently

But what is the problem to have a mail that fail?

seb-odoo replied:
What I want is:
- if the send works, I don't want the mail to be deleted
- if the send fails, I don't want an exception to be raised and I don't want an error in the log

When `force_send` is false, `raise_exception` is not read so I could remove that second arg. But the rest is necessary for the 2 cases above.

JKE-be commented on addons/website/models/website.py:
```diff
@@ -467,9 +468,64 @@ def get_current_website(self, fallback=True):
+
+        :raises: if `fallback` is True but no website at all is found
+        """
+        def _remove_port(domain_name):
```
for next time, 
_filter_domain: we avoid inner function since it is not overridable.
_remove_port: we avoid inner function for one line code called 2 times that is readable.

seb-odoo replied:
So what should I do, just create those inner function as part of the model instead? Or in some kind of tools file?

JKE-be commented:
> Probably on website helper that return (http_domain, domain) -> (scheme+(name+tld)+port, name+tld)

no ?

seb-odoo commented:
@JKE-be I applied the suggested changes (last week already, no change since then), and now I finally also fixed enterprise runbot, which was the last remaining issue. I don't plan to change anything else, you can make (hopefully final) review.

JKE-be commented:
@robodoo r+ rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward

robodoo commented:
Merged, thanks!

seb-odoo commented:
Yay!

HYROPES commented:
@seb-odoo Is that possible to push the same fix into the project.task object? 
- At moment the task still using the web.base.url despite project task do related with a website
Use case:
User can access task ticket after login the correct website portal.
However, when sharing task as link, sometimes it will direct user into wrong website as project.task only use the web.base.url
If turn on [Specific User Account] in website setting, it will be more confusing as landing on wrong website will require user register again.

- Effected version V12,13,14

seb-odoo commented:
@HYROPES I suppose it would be possible, feel free to send a support ticket describing the issue
